Add Perplexity LLM integration

vishhvak opened the pull request:
# Description

This PR introduces a new class Perplexity to the Llama Index, designed to integrate with the Perplexity AI API. The class provides a structured way to access various functionalities of the Perplexity AI models, such as completions and chat-based interactions, with support for both synchronous and asynchronous calls, as well as streaming responses.

Fixes # (issue)

Adds feature requested in #8731.

## Type of Change

Addition of an integration to an LLM provider.

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration

- [] Added new unit/integration tests
- [x] Added new notebook (that tests end-to-end)
- [x] I stared at the code and made sure it makes sense

# Suggested Checklist:

- [x] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added Google Colab support for the newly added notebooks.
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] I ran `make format; make lint` to appease the lint gods

vishhvak commented:
Added basic integration to the perplexity API, as it stands it is currently free for Pro subscribers, and offers pretty fast inference speeds for mistral, llama2, and some other models. Code is not clean/great, please do make suggestions and I shall add a notebook for documentation.

logan-markewich commented on llama_index/llms/perplexity.py:
```diff
@@ -0,0 +1,358 @@
+    @property
+    def metadata(self) -> LLMMetadata:
+        return LLMMetadata(
+            context_window=self._get_context_window(),
```
I think it might be better to also let the user override this. It can be helpful in some cases to actually shrink it (for example, our default context window for OpenAI is 3900, to help avoid token overflow errors)

If the user does not provide it, then it could default back to one of these values

vishhvak replied:
I see, that makes sense. So the right way to do this would be to make context_window also another parameter provided by the user and set it up in init - and if it is None, we set it with _get_context_window()?

logan-markewich replied:
Yup you got it 👍

vishhvak replied:
This is done, let me know if it is okay!

logan-markewich reviewed (commented):
Looks pretty good! Lets add a demo notebook to `docs/examples/llm` and link to that notebook in `docs/module_guides/models/llms/modules.md` 👍🏻

vishhvak commented on llama_index/llms/perplexity.py:
```diff
@@ -0,0 +1,358 @@
+            api_base=api_base,
+            **kwargs,
+        )
+        self.headers = {
```
Not sure what you mean - are you saying we keep headers as a parameter passed by the user in the init call?

logan-markewich replied:
Yup you got it

logan-markewich replied:
No I mean, we can set headers in super().__init__(...)

vishhvak replied:
Gotcha, set headers in init - hope i've done it correctly! :)

vishhvak commented:
> Looks pretty good! Lets add a demo notebook to `docs/examples/llm` and link to that notebook in `docs/module_guides/models/llms/modules.md` 👍🏻

@logan-markewich I've added a notebook and updated the docs as well. Also finished the updates you recommended in the comments above. Can you let me know if things are okay, or if I need to update anything else? Would like to get this merged and start using it :)

vishhvak commented:
@logan-markewich Ran a lint check as well, hopefully it works out. Might have accidentally synced my fork in between so you might see a merge commit - not sure if that was a mistake - if it was, please suggest how I should navigate it!
